### DISCUSSION TITLE
Builder returned invalid routes: should NOT have more than 1024 items

### DISCUSSION THREAD

**Discussion by JVillella** [vercel/vercel#5940] [Help]
* I'm using next.js
* I got this error after I setup a large 30k items rewrite rule in `next.config.js`

Building (`npm run build`) or just running (`npm run dev`) I don't see this problem, which is why I'm assuming it has to do with vercel.

Here's the output,
```
02:11:54.359 | ○  (Static)  automatically rendered as static HTML (uses no initial props)
02:11:54.359 | ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
02:11:54.359 | (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
02:11:54.359 | Next.js Analytics is enabled for this production build. You'll receive a Real Experience Score computed by all of your visitors.
02:11:54.457 | > boreal-web@0.1.0 postbuild /vercel/workpath0
02:11:54.457 | > next-sitemap
02:12:15.634 | Traced Next.js serverless functions for external files in: 20517.651ms
02:12:18.378 | Compressed shared serverless function files: 2743.875ms
02:12:23.040 | All serverless functions created in: 4660.887ms
02:12:23.580 | Uploading build outputs...
02:12:25.245 | Error: Builder returned invalid routes: should NOT have more than 1024 items
02:12:27.925 | Done with "package.json"

```

**Comment by JVillella** [comment]
Can confirm this seems to be (an undocumented?) restriction with the number of allowed rewrite rules within the `next.config.js` file.

**Reply by ghost** [reply]
In my case it had something to do with the limits that vercel set for Pro and Free Plan https://vercel.com/docs/concepts/limits/overview

**Comment by jessetinell** [comment]
Same here. I just changed the url structure of ≈ 1500 pages and got hit by this.

**Reply by JVillella** [reply]
I ended up just going to server side rendering instead of static gen.

**Comment by ghost** [comment]
Same issue here. fixed this problem by deleting a redirect function in my next.config.js file.

**Comment by timojokinen** [comment]
Just ran into this. Any way to work around this limitation?

**Reply by ghost** [reply]
I guess the only workaround ist to cut some redirects. There are only 2 possible solutions: 
1. You cut some redirects so that you don't run into this error (in my case I needed to adjust my next.config.js file)
2. You get yourself an Enterprise Plan from Vercel (I don't know if that even suits your purpose)
On the official Vercel Page there are Limits documented - For Free and Pro Plan the limit is set to 1024 but with an Enterprise Plan you can customize this https://vercel.com/docs/concepts/limits/overview

**Comment by scottgrunerud** [comment]
I just got something similar `Error: Builder returned invalid routes: should NOT be longer than 4096 characters` but I haven't changed my next.config or url structure ?

**Reply by chelsea-tnc** [reply]
Running into the same issue. Did you find a solution?

**Reply by amyegan** [reply]
There are a few [project configuration](https://vercel.com/docs/project-configuration) limits that you may be running into depending on your use case:
- [headers](https://vercel.com/docs/project-configuration#project-configuration/headers) can have a maximum of string length of 4,096 for the `source`, `key`, and `value`.
- [redirects](https://vercel.com/docs/project-configuration#project-configuration/redirects) and [rewrites](https://vercel.com/docs/project-configuration#project-configuration/rewrites) can have a maximum of string length of 4,096 for the `source` and `destination` values.

If you recently added a long header or query value, that could be the source of the error.

**Reply by scottgrunerud** [reply]
@chelsea-tnc I resolved it by shortening some of our headers, pretty much exactly what @amyegan said!

**Reply by chelsea-tnc** [reply]
Great, thank you both!

**Reply by kylemh** [reply]
Is still valid @amyegan ? The links don't reference character limits anymore.

**Reply by amyegan** [reply]
Yes, it's still valid. There is a 4096 character limit on static redirects, rewrites, and headers. And the limit for [Routes Created per Deployment](https://vercel.com/docs/limits/overview#routes-created-per-deployment) is 1024. But you could use [dynamic redirects](https://vercel.com/docs/edge-network/redirects#dynamic-redirects) if more are needed.


https://vercel.com/guides/how-can-i-increase-the-limit-of-redirects-or-use-dynamic-redirects-on-vercel
